packages/opentelemetry-context-zone-peer-dep/src/ZoneContextManager.ts

```diff
@@ -248,4 +248,8 @@ export class ZoneContextManager implements ContextManager {
 
     return newZone.run(fn, context);
   }
+
+  withAsync(context: Context, fn: Function) {
```

Comment:
Can you add the same async tests to this scope manager? If we add a method to the tracer which accepts an async method, users will expect it to work in both node and web.

Reply:
The test that was not working with async hooks doesn't work with ZoneJS:
```js
it('should finally restore an old scope', async () => {
      const scope1 = '1' as any;
      const scope2 = '2' as any;
      let done = false;

      await contextManager.withAsync(scope1, async () => {
        assert.strictEqual(contextManager.active(), scope1);
        await contextManager.withAsync(scope2, async () => {
          assert.strictEqual(contextManager.active(), scope2);
          done = true;
        });
        assert.strictEqual(contextManager.active(), scope1);
      });

      assert.ok(done);
    });
```

I'm guessing that's simply because zonejs can't track new async context created by V8. Should we just throw an error on the zone implementation saying that they should use `with` ?

Reply:
We cannot throw an error here. If we add `withSpanAsync` to the tracer API it must work. Similarly, if we add `withAsync` to the context API it must work. I think we may want to remove `withAsync` from the interface and have it just be on the Async Hooks context manager. Then, if users need to use it, they can call the context manager directly.

Reply:
I think that would miss user that use this API inside the browser, the API in itself works but not in specific edge cases. I believe the case above is pretty rare. WDYT ?

Reply:
It's actually an issue with Zone.js itself, see https://github.com/angular/angular/issues/31730

cc @obecny, do have any idea on how to do that ?